Refactor rule and config caching

radumereuta opened the pull request:
Partial fix for: #1867
These are incremental changes to the inner parser mostly based on refactoring and doing caching before even getting close to parsing.

I've addressed one of the issues you pointed out. The comments about `loadCaches` we're clarified on slack.

I will post some timings tomorrow.

radumereuta commented:
I hope you will ask me to review when this is ready.

cos commented:
sure

cos commented:
@andreistefanescu, please review
@dwightguth and @radumereuta, you may want to have a look
Context:
I am not happy about the result of this PR but I think it is a significant improvement over what was there before. So please review this as a delta, not judging whether the end result is very good by any means.

I spent more than a day trying to disambiguate what was there. @dwightguth and @radumereuta, if you particularly do not like something, feel free to change it yourself.

cos commented:
review after #1958

radumereuta commented on kernel/src/main/java/org/kframework/Compiler.java:
```diff
@@ -0,0 +1,42 @@
+@API
+public class Compiler {
+    /**
+     * Generates the definition containing the modules appropriate for generating rule parsers.
```
I'm having a hard time wrapping my head around this sentence.
I don't understand why you need an entire definition to generate a parser for a single module.
The signature should be `SomeKindOfParserWrapper gen(Definition baseK, Module whichModule)`.

Please explain.

cos replied:
I don't want to deal with the parser here. These are just a la carte compiler passes. See above.

cos replied:
I think you know better what `getRuleGrammar` so please feel free to improve the documentation. The documentation from the `RuleGrammarGenerator` doesn't make much sense:
```
    /**
     * Creates the seed module that can be used to parse rules.
     * Imports module markers RULE-CELLS and K found in /include/kast.k.
     * @param mod The user defined module from which to start.
     * @return a new module which imports the original user module and a set of marker modules.
     */
```

cos replied:
Why is it a "seed" module? Does the results actually parse the cells or just puts some non-functional markers? I know it's the latter, but then the name is misleading. It has mislead me at least, and, after figuring it out, I spent one day just trying to carve out of `Kompile` the code dealing with generating the appropriate syntax for parsing rules with abstract configuration.

radumereuta commented on kernel/src/main/java/org/kframework/Definition.java:
```diff
@@ -0,0 +1,87 @@
+     * The main module of the definition will be last module defined in the text file.
+     */
+    public static org.kframework.definition.Definition from(String definitionText) throws IOException {
+        Pattern pattern = Pattern.compile("module ([A-Z_]*)");
```
What? But we have parsers for things like this...
This is so error prone. This needs fixing.

cos replied:
I agree this is not optimal. Feel free to pick it up and fix it.

radumereuta commented on kernel/src/main/java/org/kframework/Definition.java:
```diff
@@ -0,0 +1,87 @@
+        File tempDir = new File(tmpDir.getAbsolutePath() + File.pathSeparator + "tempDir");
+        File definitionDir = new File(tmpDir.getAbsolutePath() + File.pathSeparator + "definitionDir");
+        File workingDir = new File(tmpDir.getAbsolutePath() + File.pathSeparator + "workingDir");
+        File kompiledDir = new File(tmpDir.getAbsolutePath() + File.pathSeparator + "kompiledDir");
```
`tmpDir` and `tempDir`. The names of variables are really important for code readability.
Maybe the first one should be `tempDirBase`?

cos replied:
running into this issue highlights a larger issue with the whole `tempDir` concept :), but anyhow... I'll rename.

cos commented:
Jenkins: test this please.

radumereuta commented on kernel/src/main/java/org/kframework/Definition.java:
```diff
@@ -0,0 +1,87 @@
+
+        org.kframework.definition.Definition definition = parserUtils.loadDefinition(
+                mainModuleName,
+                mainModuleName,
```
But the syntax module can be different, and we have a convention to determine which one it is, why aren't you handling this case in the API?

cos replied:
because a) which module we use for syntax should have nothing to do with parsing a definition object, and b) generally, I want to keep the `@API` as restrictive as possible.

radumereuta commented on kernel/src/main/java/org/kframework/Parser.java:
```diff
@@ -23,15 +25,21 @@
-    private Parser(Module module) {
-        this.parseInModule = new ParseInModule(module);
+    private Parser(Module module) throws IOException {
+        // TODO: remove hack once the frontend is cleaner; also remove the IOException once the hack is cleared
```
What's not clean in the front-end? Please be more specific.
We usually put a name on TODOs, so we know who they belong to.

cos replied:
The fact that `ParseInModule` takes three different modules as input; modules which, even after spending time refactoring that parse, I still don't know what they do.

Roughly I understand they contains different syntax for disambiguation, etc., but the clean thing would be to do all that at the KORE level, not as a black box using different data structures.

radumereuta commented on kernel/src/main/java/org/kframework/Parser.java:
```diff
@@ -23,15 +25,21 @@
+    private Parser(Module module) throws IOException {
+        // TODO: remove hack once the frontend is cleaner; also remove the IOException once the hack is cleared
+        if (module.name().endsWith(RuleGrammarGenerator.RULE_CELLS)) {
+            org.kframework.definition.Definition definitionWithBuiltins = Definition.from("require \"domians.k\"", "K");
```
To generate parsers only `kast.k` is needed. Why are you including ALL the builtins?

cos replied:
Ok. This is what I mean by the frontend not being "clean". What does the `RuleGrammarGenerator` do? Is it rule for generating the parser grammar for rules, or does it contain a mismatch of 5 other things? Actually, does it even generate anything for the parser, or does it simply prepare definitions for the parser, without having anything to do with the parser grammar? (I know definitions could be considered grammar, but that's not how we usually refer to them). Why do I need to know that only the file names `kast.k` is needed for instantiating it?

cos replied:
"clean" means that I don't need to know how the entire thing is wired together before using a feature as simple conceptually as parsing. If you want SEish terms, it also means low coupling and high cohesion, which is not a characteristic of this code.

cos replied:
I'm rambling a bit, but I'm pissed because, after we built this from scratch, I spent a few days just disentangling things. I feel like every time I'm not watching something like a hawk, it quickly becomes all mangled.

dwightguth commented:
I haven't gone over this in much detail but already I can tell you that if you are calling Definition.from anywhere in the main pipeline this is worse than before, and you simply can't do what you did with checked exceptions, it doesn't scale. If you want to bubble an exception more than one conceptual level up the call stack it needs to inherit from RuntimeException. This **cannot** be merged until you fix the issue with checked exceptions, and I don't have time to fix it for you. Sorry if you don't like that, but them's the breaks.

I will look into this deeper on another day, but it should be enough to get you started.

cos commented:
@dwightguth, I'm not a big fan of checked exceptions either, but it is the Java way. `IOException`s are actually one of the few cases where checked exceptions make some sense. Code doing IO is more fragile so you want to keep it relatively shallow and only use its results. Having many places that throw IOExceptions is usually a sign that something went wrong with the design. Our code is relatively decent in this respect.

In conclusion, I think it's best to keep the IO checked exceptions as they are after this PR.

grosu commented:
Guys, this feels like painful dejavu to me.  In the past, like 2 years ago or so, we had checked exceptions everywhere.  Everything was fine at the beginning, but as we added more and more stuff things started to fall apart, to an extent that it would take a lot of time and boilerplate code to add a new feature.  @radumereuta probably still remembers those times, @traiansf for sure.  They just grew organically into a mess that hampered modularity and code clarity.

Then we decided it was time to take the problem of exceptions seriously into our hands and we've redone the entire thing using unchecked exceptions.  This took quite a bit of time and it was a deliberate decision, motivated by a lot of pain in the past.  So if it is to learn from the past, I would say that we should stay with unchecked exceptions.

Yes, unchecked exceptions may hide program logic errors, but they give us more modularity and flexibility, that we ended up desperately needing.  Propagating checked exceptions across abstraction layers breaks abstraction by forcing the callers of an API to be concerned about the details of how the implementation of the API might fail.  That, in particular, makes developing K modules by third parties, say a group which wants to develop a push-down model checker, more painful and dependent upon our particular implementation of the API.

I guess @cos was not in the team when we faced those problems, so those who were should speak up.  In what regards me, I would rather not change the current approach unless we have a very very very strong motivation to, and a very very very strong guarantee that we will not run into the same problems that we ran into the past.

cos commented:
Again, I'm not a strong supporter of checked exceptions. I think they are, like many other concepts (injections, polymorphism, etc.), a good feature unless misused or abused. I have worked on projects where they are not an issue, but I agree they can become one if you're not careful.

In this particular case, we're talking about the API of our tool. I think it makes more sense for users to receive back a well-known `java.io.IOException` than some custom exception we make. And `IOException` is a checked exception and we cannot do anything about that.

In any case, if you feel strongly about this, I will use some unchecked exception -- not worth a very long discussion.

grosu commented:
@cos, please trust me, I would be all for checked exceptions if we hadn't gone through what we've gone in the past.  Maybe things are cleaner now and we would not run into the same problems, and maybe it is indeed a good methodology to give our users an IOException as you said, but it is a risk that I'm afraid of taking, based on prior bad experience.  This is all about risk management here, not about gut feeling nor about being right or wrong.  I really appreciate your flexibility and understanding.

cos commented:
ok, I removed the new checked `IOException` annotations, and a few old ones.

cos commented:
Jenkins: test this please

grosu commented on kernel/src/main/java/org/kframework/Parser.java:
```diff
@@ -24,14 +26,20 @@
 
     private Parser(Module module) {
-        this.parseInModule = new ParseInModule(module);
+        // TODO: remove hack once the frontend is cleaner; also remove the IOException once the hack is cleared
```
IOException?  There seems to be none here.  Also, for the sake of the future, can you very briefly write what the hack to be removed is?

grosu commented on kernel/src/main/java/org/kframework/kdep/KDepFrontEnd.java:
```diff
@@ -82,12 +84,13 @@ protected int run() {
-                        .map(files::resolveWorkingDirectory).collect(Collectors.toList()),
-                Lists.newArrayList(Kompile.BUILTIN_DIRECTORY)));
+                                .map(files::resolveWorkingDirectory).collect(Collectors.toList()),
+                        Lists.newArrayList(Kompile.BUILTIN_DIRECTORY)));
```
the indentation in the two lines above does not look right to me (as displayed by github at least).

grosu commented:
Although not asked, I've also looked through the code in this PR.  Looks good enough to me, considering the deadline.  I wished we didn't hardwire naming conventions everywhere in the code, such as the "-PROGRAM-PARSING" naming convention, to make them easy to change in the future, but that is not urgent, so let's not worry about it.  So if the others are happy with it, we should merge this asap.  Thanks again, @cos.

andreistefanescu commented on kernel/src/main/java/org/kframework/kompile/KompileFrontEnd.java:
```diff
@@ -67,7 +67,8 @@ public int run() {
 
         Kompile kompile = new Kompile(options, files, kem, sw);
         //TODO(dwightguth): handle start symbols
-        CompiledDefinition def = kompile.run(options.mainDefinitionFile(), options.mainModule(), options.syntaxModule(), Sorts.K(), koreBackend.get().steps(kompile));
```
debugging leftover?

andreistefanescu commented:
The high-level design is a step in the right direction towards a clean(er) front-end. However, I did not look over the low-level details, as they are quite hard to follow.

radumereuta committed:
upConfiguration

radumereuta committed:
Compact code.

ehildenb committed:
kernel/DefinitionParsing: factor out registerWarnings

radumereuta committed:
Factor in ruleConfigBubbles

in preparation for splitting parsing from syntax generation

radumereuta committed:
Make a separate step for cached sentences

radumereuta committed:
Split parseConfigBubbles into parse and convert

This way we can take advantage of a separate step to do caching and avoid creating extra scanners

radumereuta committed:
Split performParse in two

1. setup scanner
2. parse a bubble

radumereuta committed:
Compute extensionModule before config desugaring

radumereuta commented:
I ran `kompile -v --backend haskell` over a simple test definition (no kserver), and here are the outputs:
```
$ cat test.k
module TEST
  imports TEST-SYNTAX
  imports INT

  configuration <k> $PGM:Int </k>
  rule 0 => 1
endmodule
```
this branch first time:
```
Outer parsing [58 modules]                                   =  0.706s
New scanner: TEST-CONFIG-CELLS                               =  0.130s
New scanner: STDOUT-STREAM-CONFIG-CELLS                      =  0.205s
New scanner: STDIN-STREAM-CONFIG-CELLS                       =  0.207s
Parse configurations [3/3 declarations]                      =  0.419s
New scanner: TEST-RULE-CELLS                                 =  0.124s
New scanner: STDIN-STREAM-RULE-CELLS                         =  0.190s
New scanner: STDOUT-STREAM-RULE-CELLS                        =  0.185s
Parse rules [101/101 rules]                                  =  0.742s
Validate definition                                          =  0.071s
Apply compile pipeline                                       =  0.248s
Kompile to kore                                              =  0.002s
Save to disk                                                 =  0.021s
Backend                                                      =  0.288s
Cleanup                                                      =  0.000s
Total                                                        =  2.507s
```
this branch second time:
```
Outer parsing [58 modules]                                   =  0.711s
Parse configurations [0/3 declarations]                      =  0.262s
Parse rules [0/101 rules]                                    =  0.072s
Validate definition                                          =  0.084s
Apply compile pipeline                                       =  0.319s
Kompile to kore                                              =  0.002s
Save to disk                                                 =  0.034s
Backend                                                      =  0.335s
Cleanup                                                      =  0.000s
Total                                                        =  1.829s
```
master branch first time:
```
Outer parsing [57 modules]                                   =  0.736s
New scanner: TEST-CONFIG-CELLS                               =  0.126s
New scanner: STDOUT-STREAM-CONFIG-CELLS                      =  0.173s
New scanner: STDIN-STREAM-CONFIG-CELLS                       =  0.181s
Parse configurations [3/3 declarations]                      =  0.802s
New scanner: TEST-RULE-CELLS                                 =  0.123s
New scanner: STRING-COMMON-RULE-CELLS                        =  0.177s
New scanner: STRING-KORE-RULE-CELLS                          =  0.182s
New scanner: K-IO-RULE-CELLS                                 =  0.248s
New scanner: STDOUT-STREAM-RULE-CELLS                        =  0.252s
New scanner: STDIN-STREAM-RULE-CELLS                         =  0.258s
Parse rules [101/101 rules]                                  =  0.699s
Validate definition                                          =  0.063s
Apply compile pipeline                                       =  0.255s
Kompile to kore                                              =  0.001s
Save to disk                                                 =  0.032s
Backend                                                      =  0.291s
Cleanup                                                      =  0.001s
Total                                                        =  2.890s
```
master branch second time:
```
Outer parsing [57 modules]                                   =  0.714s
New scanner: TEST-CONFIG-CELLS                               =  0.127s
New scanner: STDOUT-STREAM-CONFIG-CELLS                      =  0.176s
New scanner: STDIN-STREAM-CONFIG-CELLS                       =  0.185s
Parse configurations [0/3 declarations]                      =  0.738s
New scanner: TEST-RULE-CELLS                                 =  0.119s
New scanner: STRING-COMMON-RULE-CELLS                        =  0.137s
New scanner: STRING-KORE-RULE-CELLS                          =  0.136s
New scanner: K-IO-RULE-CELLS                                 =  0.201s
New scanner: STDOUT-STREAM-RULE-CELLS                        =  0.200s
New scanner: STDIN-STREAM-RULE-CELLS                         =  0.216s
Parse rules [0/101 rules]                                    =  0.473s
Validate definition                                          =  0.081s
Apply compile pipeline                                       =  0.275s
Kompile to kore                                              =  0.001s
Save to disk                                                 =  0.029s
Backend                                                      =  0.315s
Cleanup                                                      =  0.001s
Total                                                        =  2.637s
```

master: make -j 32 regression-new -> 1m52s
current: make -j 32 regression-new -> 1m45s
With the kserver on.

master: make -j 32 regression-new -> 7m16s
current: make -j 32 regression-new -> 7m10s
With the kserver off.

radumereuta commented:
@dwightguth  I figured out why Algorand was failing.
On master the parsing of configurations is actually done in sequence in a transformer. And right after parsing you also do configuration expansion. Algorand has multiple modules with configurations. It parses configuration from the first module and generates all the config syntax based on MAP. And it parses fine because it doesn't use .Map inside the configuration. Then when it comes to the module with problems it passes.
See my minimal example here: https://github.com/kframework/k/pull/1965#issuecomment-840655317
On this branch, I've split these two steps. This will allow me to make config parsing parallel. But that's going to come in the next step.

radumereuta commented:
Here are some timings for larger definitions:
EVM: master 1m06s, this branch: 1m07s
Algorand: master: 2m26s, this branch 2m22s
C: master 2m11s, this branch 2m14s

dwightguth commented on kernel/src/main/java/org/kframework/kompile/DefinitionParsing.java:
```diff
@@ -277,63 +279,73 @@ protected Definition resolveConfigBubbles(Definition definition, Module defaultC
+
+            ParseCache cache = loadCache(ruleParserModule);
+            try (ParseInModule parser = RuleGrammarGenerator.getCombinedGrammar(cache.getModule(), isStrict, profileRules, files)) {
+                if (stream(m.localSentences()).filter(s -> s instanceof Bubble && ((Bubble) s).sentenceType().equals(configuration))
```
We don't need this condition anymore, do we? We know there's at least one bubble that wasn't in the cache because we called `resolveCachedBubbles` and we still have configuration bubbles afterwards.

dwightguth reviewed (changes requested):
I don't see how any of my prior concerns were addressed here. We are still constructing multiple ParseInModules and it's still not clear what the difference is between performParse and parseBubble and why there are both.

radumereuta committed:
Remove redundant check

radumereuta committed:
Remove performParse

Embed the Scanner directly in ParseInModule before it reaches the parser

radumereuta commented:
I've removed `performParse`. I want `ParseInModule` to be finished before we send it to the parser (I find it cleaner). This is why I have added `setScanner`.

I'm not sure why you are saying that I am building multiple `ParseInModule`. Can you be more precise?

This is an intermediate state of the PR, which doesn't include the minimum number of scanners generated.
This way we don't break any of the definitions, but we are still getting the benefit of not building scanners when sentences were cached.

The next change I want to add is to parse Configurations in parallel. With the same number of scanners for now. But I want to make sure the PR is OK for now.

radumereuta commented:
@dwightguth please review.
I've fixed all your comments (at least the ones I could find in the old PRs).
I've also included parallel parsing for configurations, this time generating one scanner for one parser.
For rules, I'm creating a minimal list of scanners depending on module imports and reusing them every time it's possible. All the definitions I can find generate at most 3 scanners for programs. Main module and two for streams.

I've also tested the major semantics and integration tests and I can say that the memory usage stays the same, and timings are equal or slightly better (see my previous comment with updated timings).

For the second compilation then the difference is really noticeable.
Here is EVM:
first time:
```
Outer parsing [76 modules]                                   =  0.908s
New scanner: STDOUT-STREAM-CONFIG-CELLS                      =  0.217s
New scanner: STDIN-STREAM-CONFIG-CELLS                       =  0.210s
New scanner: EVM-CONFIG-CELLS                                =  0.505s
Parse configurations [3/3 declarations]                      =  1.173s
New scanner: STDIN-STREAM-RULE-CELLS                         =  0.272s
New scanner: ETHEREUM-SIMULATION-RULE-CELLS                  =  1.129s
New scanner: STDOUT-STREAM-RULE-CELLS                        =  0.203s
Parse rules [1479/1479 rules]                                = 27.917s
Validate definition                                          =  0.211s
Apply compile pipeline                                       =  5.953s
Kompile to kore                                              =  3.183s
Save to disk                                                 =  0.148s
Backend                                                      =  0.000s
Cleanup                                                      =  0.000s
Total                                                        = 39.507s
```
second time:
```
Outer parsing [76 modules]                                   =  0.837s
Parse configurations [0/3 declarations]                      =  0.501s
Parse rules [0/1479 rules]                                   =  0.196s
Validate definition                                          =  0.272s
Apply compile pipeline                                       =  6.052s
Kompile to kore                                              =  3.274s
Save to disk                                                 =  0.146s
Backend                                                      =  0.001s
Cleanup                                                      =  0.000s
Total                                                        = 11.291s
```

dwightguth commented:
I feel like we're going in circles here. We were almost ready to merge this, I had a couple minor suggestions to erase some places where the performance looked like it was regressing, and now we're back to the code being completely different from what it was when I reviewed last. Why can't we just take the code that I was close to approving, modify it so it's not constructing multiple ParseInModules, clarify the conceptual difference between the two functions I was confused about earlier, and get this merged?

radumereuta commented:
You asked me to make incremental changes so you can review commits independently. The last feedback you gave me was:
>I don't see how any of my prior concerns were addressed here. We are still constructing multiple ParseInModules and it's still not clear what the difference is between performParse and parseBubble and why there are both.

which I fixed with `Remove redundant check` and `Remove performParse`

Then we talked about parsing configurations and needing scanners for every module. And I added these changes: `Parse configs in parallel`

The only thing which is new is `Rule parser uses minimal amount of scanners`. If you want I can temporarily remove it, but this final commit is the actual reason I started these changes. So I don't want to merge into mater without it.

radumereuta committed:
Remove setScanner and make it a constructor parameter

rv-jenkins committed:
Merge branch 'master' into refactorRuleParsing7
